File: matiasmasso/Mat.NET/DT2/DTOWebMenu.vb
Public Class DTOWebMenu
    Property Groups As List(Of DTOWebMenuGroup)
End Class

Public Class DTOWebMenuGroup
    Inherits DTOBaseGuid

    Property Ord As Integer
    Property LangText As DTOLangText

    Property [Private] As Boolean
    Property Items As List(Of DTOWebMenuItem)

    Property Url As String

    Public Enum wellknowns
        NotSet
        Langs
        Products
    End Enum

    Public Sub New()
        MyBase.New()
        _Items = New List(Of DTOWebMenuItem)
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
        _Items = New List(Of DTOWebMenuItem)
    End Sub

    Shared Function wellknown(value As DTOWebMenuGroup.wellknowns) As Guid
        Dim retval As Guid = Nothing
        Select Case value
            Case DTOWebMenuGroup.wellknowns.Langs
                retval = New Guid("69D3D691-8F8E-41C9-8711-3E8A189EF113")
            Case DTOWebMenuGroup.wellknowns.Products
                retval = New Guid("81F161B9-4513-4388-8CB6-329E6005ECE4")
        End Select
        Return retval
    End Function

    Shared Function Nom(oWebMenuGroup As DTOWebMenuGroup, oLang As DTOLang) As String
        Dim retval As String = oWebMenuGroup.LangText.tradueix(oLang)
        Return retval
    End Function

    Shared Function LangsGroup(oLang As DTOLang) As DTOWebMenuGroup
        Dim retval As New DTOWebMenuGroup(DTOWebMenuGroup.wellknown(DTOWebMenuGroup.wellknowns.Langs))
        retval.LangText = New DTOLangText("Idioma", "Llengua", "Language", "idioma")


        DTOWebMenuGroup.AddItem(retval, "Español", "Espanyol", "Spanish", "Espanhol", "/esp", oLang.Equals(DTOLang.ESP))
        DTOWebMenuGroup.AddItem(retval, "Catalán", "Català", "Catalan", "Catalão", "/cat", oLang.Equals(DTOLang.CAT))
        DTOWebMenuGroup.AddItem(retval, "Inglés", "Anglés", "English", "Inglês", "/eng", oLang.Equals(DTOLang.ENG))
        DTOWebMenuGroup.AddItem(retval, "Portugués", "Portuguès", "Portuguese", "Português", "/por", oLang.Equals(DTOLang.POR))
        Return retval
    End Function


    Shared Function SorteosGroup() As DTOWebMenuGroup
        Dim retval As New DTOWebMenuGroup
        retval.LangText = New DTOLangText("Sorteos", "Sortejos", "Raffles", "Sorteios")
        retval.Url = "/sorteos"
        Return retval
    End Function

    Shared Function NoticiasGroup() As DTOWebMenuGroup
        Dim retval As New DTOWebMenuGroup
        retval.LangText = New DTOLangText("Noticias", "Noticies", "News", "Notícias")
        retval.Url = "/news"
        Return retval
    End Function

    Shared Function SignUpMenuGroup() As DTOWebMenuGroup
        Dim retval As New DTOWebMenuGroup
        retval.LangText = New DTOLangText("Registrarse", "Registrar-se", "Sign up", "Inscrever-se")
        retval.Url = "/registro"
        Return retval
    End Function

    Shared Function LoginGroup() As DTOWebMenuGroup
        Dim retval As New DTOWebMenuGroup()
        retval.LangText = New DTOLangText("Acceder", "Accedir", "Log in", "Entrar")
        retval.Url = "/pro"
        Return retval
    End Function

    Shared Function AddItem(oParent As DTOWebMenuGroup, sEsp As String, sCat As String, sEng As String, sPor As String, Optional sNavigateTo As String = "#", Optional BlSelected As Boolean = False) As DTOWebMenuItem
        Dim retval As New DTOWebMenuItem()
        With retval
            .LangText = New DTOLangText(sEsp, sCat, sEng, sPor)
            .Url = sNavigateTo
            .Actiu = BlSelected
        End With
        oParent.Items.Add(retval)
        Return retval
    End Function

End Class

Public Class DTOWebMenuItem
    Inherits DTOBaseGuid

    Property LangText As DTOLangText
    Property Ord As String
    Property Url As String
    Property Actiu As Boolean
    Property Group As DTOWebMenuGroup
    Property Rols As List(Of DTORol)


    Public Sub New()
        MyBase.New()
        _Rols = New List(Of DTORol)
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
        _Rols = New List(Of DTORol)
    End Sub

    Shared Function Nom(oWebMenuItem As DTOWebMenuItem, oLang As DTOLang) As String
        Dim retval As String = oWebMenuItem.LangText.tradueix(oLang)
        Return retval
    End Function
End Class

File: matiasmasso/Mat.NET/DTO/DTOWebMenu.cs
using System;
using System.Collections.Generic;

namespace DTO
{
    public class DTOWebMenu
    {
        public List<DTOWebMenuGroup> Groups { get; set; }
    }

    public class DTOWebMenuGroup : DTOBaseGuid
    {
        public int Ord { get; set; }
        public DTOLangText LangText { get; set; }

        //public bool Private { get; set; }
        public List<DTOWebMenuItem> Items { get; set; }

        //public string Url { get; set; }
        public DTOLangText LangUrl { get; set; }

        public enum Wellknowns
        {
            NotSet,
            Langs,
            Products,
            Queries,
            Forms,
            Perfil,
            Developer
        }

        public DTOWebMenuGroup() : base()
        {
            Items = new List<DTOWebMenuItem>();
        }

        public DTOWebMenuGroup(Guid oGuid) : base(oGuid)
        {
            Items = new List<DTOWebMenuItem>();
        }

        public static Guid Wellknown(DTOWebMenuGroup.Wellknowns value)
        {
            Guid retval = default(Guid);
            switch (value)
            {
                case DTOWebMenuGroup.Wellknowns.Langs:
                    {
                        retval = new Guid("69D3D691-8F8E-41C9-8711-3E8A189EF113");
                        break;
                    }

                case DTOWebMenuGroup.Wellknowns.Products:
                    {
                        retval = new Guid("81F161B9-4513-4388-8CB6-329E6005ECE4");
                        break;
                    }
                case DTOWebMenuGroup.Wellknowns.Perfil:
                    {
                        retval = new Guid("F10C64A3-840C-490C-9FDE-953EE19A6B51");
                        break;
                    }
            }
            return retval;
        }

        public static string Nom(DTOWebMenuGroup oWebMenuGroup, DTOLang oLang)
        {
            string retval = oWebMenuGroup.LangText.Tradueix(oLang);
            return retval;
        }

        public static DTOWebMenuGroup LangsGroup(DTOLang oLang)
        {
            DTOWebMenuGroup retval = new DTOWebMenuGroup(DTOWebMenuGroup.Wellknown(DTOWebMenuGroup.Wellknowns.Langs));
            retval.LangText = new DTOLangText("Idioma", "Llengua", "Language", "idioma");


            DTOWebMenuGroup.AddItem(retval, "Español", "Espanyol", "Spanish", "Espanhol", "/esp", oLang.Equals(DTOLang.ESP()));
            DTOWebMenuGroup.AddItem(retval, "Catalán", "Català", "Catalan", "Catalão", "/cat", oLang.Equals(DTOLang.CAT()));
            DTOWebMenuGroup.AddItem(retval, "Inglés", "Anglés", "English", "Inglês", "/eng", oLang.Equals(DTOLang.ENG()));
            DTOWebMenuGroup.AddItem(retval, "Portugués", "Portuguès", "Portuguese", "Português", "/por", oLang.Equals(DTOLang.POR()));
            return retval;
        }


        public static DTOWebMenuGroup SorteosGroup()
        {
            DTOWebMenuGroup retval = new DTOWebMenuGroup();
            retval.LangText = new DTOLangText("Sorteos", "Sortejos", "Raffles", "Sorteios");
            retval.LangUrl = new DTOLangText("/sorteos","/sortejos","/raffles","/sorteios");
            return retval;
        }

        public static DTOWebMenuGroup NoticiasGroup()
        {
            DTOWebMenuGroup retval = new DTOWebMenuGroup();
            retval.LangText = new DTOLangText("Noticias", "Notícies", "News", "Notícias");
            retval.LangUrl = new DTOLangText("/news");
            return retval;
        }

        public static DTOWebMenuGroup SignUpMenuGroup()
        {
            DTOWebMenuGroup retval = new DTOWebMenuGroup();
            retval.LangText = new DTOLangText("Registrarse", "Registrar-se", "Sign up", "Inscrever-se");
            retval.LangUrl = new DTOLangText("/registro");
            return retval;
        }

        public static DTOWebMenuGroup LoginGroup()
        {
            DTOWebMenuGroup retval = new DTOWebMenuGroup();
            retval.LangText = new DTOLangText("Acceder", "Accedir", "Log in", "Entrar");
            retval.LangUrl = new DTOLangText("/pro");
            return retval;
        }

        public static DTOWebMenuItem AddItem(DTOWebMenuGroup oParent, string sEsp, string sCat, string sEng, string sPor, string sNavigateTo = "#", bool BlSelected = false)
        {
            DTOWebMenuItem retval = new DTOWebMenuItem();
            {
                var withBlock = retval;
                withBlock.LangText = new DTOLangText(sEsp, sCat, sEng, sPor);
                withBlock.LangUrl = new DTOLangText(sNavigateTo);
                withBlock.Actiu = BlSelected;
            }
            oParent.Items.Add(retval);
            return retval;
        }
    }

    public class DTOWebMenuItem : DTOBaseGuid
    {
        public DTOLangText LangText { get; set; }
        public string Ord { get; set; }
        //public string Url { get; set; }
        public DTOLangText LangUrl { get; set; }
        public bool Actiu { get; set; }
        public DTOWebMenuGroup Group { get; set; }
        public List<DTORol> Rols { get; set; }


        public DTOWebMenuItem() : base()
        {
            Rols = new List<DTORol>();
        }

        public DTOWebMenuItem(Guid oGuid) : base(oGuid)
        {
            Rols = new List<DTORol>();
        }

        public static string Nom(DTOWebMenuItem oWebMenuItem, DTOLang oLang)
        {
            string retval = oWebMenuItem.LangText.Tradueix(oLang);
            return retval;
        }
    }
}
